lib/saluki-core/src/components/renderers/mod.rs

```diff
@@ -0,0 +1,36 @@
+//! Renderer component basics.
+
+use async_trait::async_trait;
+use saluki_error::GenericError;
+
+mod builder;
+pub use self::builder::RendererBuilder;
+
+mod context;
+pub use self::context::RendererContext;
+
+/// A renderer.
+///
+/// Renderers are the bridge between forwarders and the rest of the topology. They are responsible for rendering
+/// telemetry events into output payloads that can then be forwarded. Most renderers are specific to a particular system
+/// and not simply equivalent to a certain encoding or serialization format: while two renderers may both produce JSON,
+/// they may produce different JSON formats such that one format only works for product A and the other only works for
+/// product B. In essence, the process of taking telemetry events and sending them to product A ends up becoming the sum
+/// of a specific rendered and forwarder combination.
+///
+/// Examples of typical renderers include Datadog Metrics, Events, and Service Checks.
+#[async_trait]
+pub trait Renderer {
+    /// Runs the renderer.
+    ///
+    /// The renderer context provides access primarily to the event stream, used to receive events sent to the renderer,
+    /// and the forwarder, used to send payloads to the downstream forwarder in the topology, as well as other
+    /// information such as the component context.
+    ///
+    /// Renderers are expected to run indefinitely until their event stream is terminated, or an error occurs.
+    ///
+    /// # Errors
+    ///
+    /// If an unrecoverable error occurs while running, an error is returned.
+    async fn run(self: Box<Self>, context: RendererContext) -> Result<(), GenericError>;
```

Comment:
Related, this would be an interesting one to try to scope down the API a lot more. Essentially have individual implementations focus only on the actual logic, and wrap it with some kind of driver that is the spawned task. So this trait could be more like `add_event`, `flush`, `tick`, etc, all synchronous if possible.

Could also be something to explore later, once the concept is more established.

Reply:
Hmm, maybe.

It would definitely be possible, although I wonder if we can properly encode all necessary behaviors with just the individual operations (`add_event`, `flush`, etc) and a generic driver on top.

One thought is that we do that with a default implementation of `run`. Basically just have it process all ready events (via `add_event`) and then `flush` at the end before looping all over again. If code needed a custom implementation of `run` -- specialized flushing behavior, etc -- they could just override `run` in their implementation, or create a wrapper `Renderer` impl or something.

Reply:
Or, alternatively, maybe we leave the exact trait as-is but add something equivalent to `SynchronousTransform`, and _that's_ where we implement the model you mention.. and perhaps we end up with like two actual implementations of `Renderer`, like one for "add and flush as soon as possible" and one for "configurable batching/buffering/dwell time"... but all of the interesting implementations are all of `SynchronousRenderer` or whatever.

Reply:
I ended up switching to having both `Renderer` and adding a new `IncrementalRenderer` trait, which defines a method for processing a single event, and a method for flushing. Processing the event gives back a result that indicates if processing can proceed (or if a flush must be triggered) which I think could be a good middle ground for the moment.

This will let us implement the actual renderers in terms of how to process each event and flush, while allowing us to decouple all of the other thorny parts like buffering, shutdown, etc.